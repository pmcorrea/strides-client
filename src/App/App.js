import React, { Component } from "react";
import { Route, Link, Redirect } from "react-router-dom";

import './App.css';
import '../Components/Home/home.css';

import MainContext from "../Contexts/MainContext";
import TokenHelpers from "../Services/token-helpers"

import Login from "../Components/Login/login.js";
import Home from "../Components/Home/home.js";
import Register from "../Components/Register/register.js";
import AddHabit from "../Components/Add-Habit/add-habit.js";

import Profile from "../Components/Profile/profile.js";
import Habit from "../Components/Habit/habit.js";
import EditHabit from "../Components/Edit-Habit/edit-habit.js";
import Settings from "../Components/Settings/settings.js";

import HomeIcon from "../Assets/home-icon.svg"
import PlusIcon from "../Assets/plus-icon.svg"
import ProfileIcon from "../Assets/profile-icon.svg"

import { ApolloProvider, ApolloClient, createHttpLink, InMemoryCache } from '@apollo/client';
import { setContext } from "apollo-link-context";
import config from '../config'

const authLink = setContext((_, { headers }) => {
  const token = TokenHelpers.getAuthToken()
  return {
    headers: {
      ...headers,
      authorization: token ? `bearer ${token}` : null,
    }
  }
})

const httpLink = new createHttpLink({ uri: `${config.API_ENDPOINT}` })


const apolloClient = new ApolloClient({
  cache: new InMemoryCache(),
  link: authLink.concat(httpLink),
  credentials: 'include'
})


export default class App extends Component {
  constructor() {
    super()
    // this.state = {
    //   token: TokenHelpers.getAuthToken(),
      
    //   setToken: () => {
    //     let someToken = TokenHelpers.getAuthToken()
        
    //     this.setState({
    //       token: someToken
    //     })
    //   }
    // }
  }

  renderMainRoutes() {
    return (
      <>
        <Route
          path="/"
          exact
          render={routeProps => {
            return <Login {...routeProps} />
          }}
        />

        <Route
          path="/register"
          render={routeProps => {
            return <Register {...routeProps} />
          }}
        />

        <Route
          path="/home"
          render={routeProps => {
            if (!TokenHelpers.hasAuthToken()) {
              return <Redirect to="/" />
            }
            return <Home {...routeProps} />
          }}
        />

        <Route
          path="/add-habit"
          render={routeProps => {
            if (!TokenHelpers.hasAuthToken()) {
              return <Redirect to="/" />
            }
            return <AddHabit {...routeProps} />
          }}
        />

        <Route
          path="/profile"
          render={routeProps => {
            if (!TokenHelpers.hasAuthToken()) {
              return <Redirect to="/" />
            }
            return <Profile {...routeProps} />
          }}
        />

        <Route
          path="/habit/:id"
          render={routeProps => {
            if (!TokenHelpers.hasAuthToken()) {
              return <Redirect to="/" />
            }
            return <Habit {...routeProps} />
          }}
        />

        <Route
          path="/edit-habit/:id"
          render={routeProps => {
            if (!TokenHelpers.hasAuthToken()) {
              return <Redirect to="/" />
            }
            return <EditHabit {...routeProps} />
          }}
        />

        <Route
          path="/settings"
          render={routeProps => {
            if (!TokenHelpers.hasAuthToken()) {
              return <Redirect to="/" />
            }
            return <Settings {...routeProps} />
          }}
        />
      </>
    )
  }

  renderBottomBar() {
    return (
      <>
        <div className="bottom-toolbar">
          <button>
            <Link to="/home">
              <img src={HomeIcon} alt="" className="icon-bar-icon" />
            </Link>
          </button>

          <button> 
            <Link to="/add-habit">
              <img src={PlusIcon} alt="" className="icon-bar-icon" />
            </Link>
          </button>

          <button>
            <Link to="/profile">
             <img src={ProfileIcon} alt="" className="icon-bar-icon" />
            </Link>
          </button>
        </div>	
      </>
    )
  }

  render() {
    return (
      <ApolloProvider client={apolloClient}>
      <MainContext.Provider value={this.state}>
        <div className="App">
          {this.renderMainRoutes()}

          {/* {window.location.pathname === "/" | window.location.pathname === "/register"
            ? ""
            : this.renderBottomBar()
          }   */}
        </div>
      </MainContext.Provider>
      </ApolloProvider>
    );
  }
}

