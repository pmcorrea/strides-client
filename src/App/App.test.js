import React from 'react';
import ReactDOM from 'react-dom'
import { BrowserRouter } from 'react-router-dom'
import { render, cleanup } from '@testing-library/react';
import App from './App';

afterEach(cleanup)

// Snapshot Test
it ('should take a snapshot', () => {
  const { asFragment } = render(
    <BrowserRouter>
      <App />
    </BrowserRouter>
    )

  expect(asFragment(
  <BrowserRouter>
    <App />
  </BrowserRouter>)).toMatchSnapshot()
})

// Smoke Test
it('renders without crashing', () => {
  const div = document.createElement('div')

  ReactDOM.render(
    <BrowserRouter>
      <App />
    </BrowserRouter>,
    div
  )
  ReactDOM.unmountComponentAtNode(div)
})