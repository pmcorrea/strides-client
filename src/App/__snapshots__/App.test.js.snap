// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should take a snapshot 1`] = `
<DocumentFragment>
  <div
    class="App"
  >
    <div
      class="main-container"
    >
      <div
        class="login-screen"
      >
        <div
          class="login-container"
          data-testid="login-test-container"
        >
          <h1
            data-testid="login-h1-test-container"
          >
            Strides
          </h1>
          <form
            action=""
            class="login-form"
          >
            <div
              class="input-container"
            >
              <input
                name="username"
                placeholder="username"
                type="text"
              />
              <label
                for="username"
              >
                username
              </label>
            </div>
            <div
              class="input-container"
            >
              <input
                name="password"
                placeholder="username"
                type="password"
              />
              <label
                for="password"
              >
                password
              </label>
            </div>
            <div>
              <button
                type="submit"
              >
                Login
              </button>
              <a
                href="/register"
              >
                Register
              </a>
            </div>
            <div>
              <button
                type="submit"
              >
                Demo &gt;
              </button>
            </div>
          </form>
          <div
            class="login-validation"
            style="display: none;"
          />
        </div>
        <div
          class="rightside"
        >
          <img
            alt="background-img"
            src="right-container.jpg"
          />
          <div
            class="nav-links"
          >
            <a
              href="#about"
            >
              About
            </a>
            <a
              href="https://petercorrea.com"
              rel="noopener noreferrer"
              target="_blank"
            >
              Portfolio
            </a>
          </div>
          <h1
            class="title-right"
          >
            Your companion for better habits.
          </h1>
        </div>
      </div>
      <div
        class="secondary-container"
      >
        <div
          class="row-1"
        >
          <div
            class="left"
          >
            <img
              alt="mock-iphone"
              class="mock-iphone-img"
              src="IMG_4768_iphonexspacegrey_portrait.png"
            />
          </div>
          <div
            class="right"
          >
            <div>
              <h1>
                Be a
              </h1>
              <h1>
                better self
              </h1>
            </div>
            <div
              class="bullets"
            >
              <p>
                A minimalistic habit tracker
              </p>
              <p>
                Simply pick the dates and start logging.
              </p>
              <p>
                Get around to your goals and aim for perfect streaks!
              </p>
            </div>
          </div>
        </div>
        <div
          class="row-2"
        >
          <div
            class="left"
          >
            <div>
              <h1>
                Build the
              </h1>
              <h1>
                perfect routne
              </h1>
            </div>
            <div
              class="bullets"
            >
              <p>
                Track unlimited habits to build consistant routines.
              </p>
              <p>
                Plan around your time with customized schedules.
              </p>
              <p>
                Automatically generate stats and determine perfect streaks. 
              </p>
            </div>
          </div>
          <div
            class="right"
          >
            <img
              alt="mock-iphone"
              class="mock-iphone-img"
              src="IMG_4767_iphonexspacegrey_portrait.png"
            />
          </div>
        </div>
        <div
          class="row-3"
        >
          <div
            class="left"
          >
            <img
              alt="mock-iphone"
              class="mock-iphone-img"
              src="IMG_4769_iphonexspacegrey_portrait.png"
            />
          </div>
          <div
            class="right"
          >
            <h1
              id="about"
            >
              About Strides
            </h1>
            <p>
              Strides was developed using React, NodeJS, GraphQL and Postgres. Deployment tools included Vercel and Heroku. Other notable tools and libraries include: React Hooks, Router and Context, Apollo, and Express. Testing was faciliated using Testing Library, Chai, and Mocha. Security considerations included implementing a JWT authorization system with expiring tokens, hashing/salting passwords with bcrypt, sanitizing input with XSS, and CORS for secured headers.
            </p>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
